gcore/src/msg.rs

```diff
@@ -38,46 +41,76 @@ mod sys {
             data_len: u32,
             value_ptr: *const u8,
             message_id_ptr: *mut u8,
-        );
-        pub fn gr_reply_commit(message_id_ptr: *mut u8, value_ptr: *const u8);
-        pub fn gr_reply_push(data_ptr: *const u8, data_len: u32);
+        ) -> ErrorCode;
+        pub fn gr_reply_commit(message_id_ptr: *mut u8, value_ptr: *const u8) -> ErrorCode;
+        pub fn gr_reply_push(data_ptr: *const u8, data_len: u32) -> ErrorCode;
         pub fn gr_reply_to(dest: *mut u8);
         pub fn gr_send(
             program: *const u8,
             data_ptr: *const u8,
             data_len: u32,
             value_ptr: *const u8,
             message_id_ptr: *mut u8,
-        );
+        ) -> ErrorCode;
         pub fn gr_send_wgas(
             program: *const u8,
             data_ptr: *const u8,
             data_len: u32,
             gas_limit: u64,
             value_ptr: *const u8,
             message_id_ptr: *mut u8,
-        );
+        ) -> ErrorCode;
         pub fn gr_send_commit(
             handle: u32,
             message_id_ptr: *mut u8,
             program: *const u8,
             value_ptr: *const u8,
-        );
+        ) -> ErrorCode;
         pub fn gr_send_commit_wgas(
             handle: u32,
             message_id_ptr: *mut u8,
             program: *const u8,
             gas_limit: u64,
             value_ptr: *const u8,
-        );
-        pub fn gr_send_init() -> u32;
-        pub fn gr_send_push(handle: u32, data_ptr: *const u8, data_len: u32);
+        ) -> ErrorCode;
+        pub fn gr_send_init(handle: *mut u32) -> ErrorCode;
+        pub fn gr_send_push(handle: u32, data_ptr: *const u8, data_len: u32) -> ErrorCode;
         pub fn gr_size() -> u32;
         pub fn gr_source(program: *mut u8);
         pub fn gr_value(val: *mut u8);
     }
 }
 
+#[repr(transparent)]
+#[derive(Debug, Copy, Clone)]
+pub struct ErrorCode(i32);
+
+impl ErrorCode {
+    fn into_send_error(self) -> Result<(), SendError> {
+        if self.0 == 0 {
+            Ok(())
+        } else {
+            Err(SendError(()))
+        }
+    }
+
+    fn into_reply_error(self) -> Result<(), ReplyError> {
+        if self.0 == 0 {
+            Ok(())
+        } else {
+            Err(ReplyError(()))
+        }
+    }
+}
+
+/// An error occurred during sending a message
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SendError(());
+
+/// An error occurred during replying to a message
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ReplyError(());
```

Comment:
```suggestion
pub struct ReplyError;
```

Reply:
This better to resolve

Reply:
Union is used to forbid creation of structure outside of `gcore`

Reply:
Didn't catch it at the start

Brilliant move